r/src/io.cpp

```diff
@@ -178,4 +181,164 @@ void io___BufferOutputStream__Write(
+      // buffer), check that we have read enough characters to get into
+      // `src` (after which the loop below will error for invalid characters).
+      int64_t bytes_read_in = in_buf - pending_;
+      if (bytes_read_in < n_pending_) {
+        return StatusInvalidInput();
+      }
+
+      int64_t bytes_read_out = out_buf - dest->data();
+      out_bytes_used += bytes_read_out;
+
+      int64_t chars_read_in = n_pending_ + n_src_bytes_in_pending - in_bytes_left;
+      in_buf = src->data() + chars_read_in - n_pending_;
+      in_bytes_left = src->size() + n_pending_ - chars_read_in;
+    } else {
+      in_buf = src->data();
+      in_bytes_left = src->size();
+    }
+
+    // in_bytes_left >= 4 assumes a maximum of 4 bytes per character in the
+    // input (the maximum for UTF-8, UTF-16, and UTF-32). If we
+    // have more than this, it means the output buffer wasn't big enough
+    // and the next iteration of the loop will try iconv() again with a
+    // bigger buffer. When we have less than 4 bytes left, the bytes
+    // will be copied to pending_ and processed in the next call (the
+    // TransformInputStream always finishes with a src that is 0 bytes
+    // for this purpose.)
+    while (in_bytes_left >= 4) {
+      // When this is true, we will (almost) always need a new buffer
+      if (out_bytes_left < in_bytes_left) {
+        RETURN_NOT_OK(dest->Resize(dest->size() * 1.2));
+        out_buf = dest->mutable_data() + out_bytes_used;
+        out_bytes_left = dest->size() - out_bytes_used;
+      }
+
+      // iconv() can return an error code ((size_t) -1) but it's not
+      // useful as it can occur because of invalid input, because
+      // of a full output buffer, or because there are partial characters
+      // at the end of the input buffer that were not completely decoded.
+      // We handle each of these cases separately based on the number of bytes
+      // read or written.
+      uint8_t* out_buf_before = out_buf;
+
+      iconv_->iconv(&in_buf, &in_bytes_left, &out_buf, &out_bytes_left);
+
+      int64_t bytes_read_out = out_buf - out_buf_before;
+
+      if (bytes_read_out == 0) {
+        return StatusInvalidInput();
+      }
+
+      out_bytes_used += bytes_read_out;
+    }
+
+    // Keep the leftover characters until the next call to the function
+    n_pending_ = in_bytes_left;
+    if (in_bytes_left > 0) {
+      memcpy(pending_, in_buf, in_bytes_left);
+    }
+
+    // Shrink the output buffer to only the size used
+    RETURN_NOT_OK(dest->Resize(out_bytes_used, false));
+    return std::move(dest);
+  }
+
+ protected:
+  std::string from_;
+  std::shared_ptr<RIconvWrapper> iconv_;
+  uint8_t pending_[8];
+  int64_t n_pending_;
+
+  arrow::Status StatusInvalidInput() {
+    std::stringstream stream;
+    stream << "Encountered invalid input bytes ";
+    stream << "(input encoding was '" << from_ << "')";
+    return arrow::Status::IOError(stream.str());
```

Comment:
Nit: you don't need to use `StringStream` explicit, the Status factory methods will do it for you, e.g.:
```c++
return Status::IOError(
    "Encountered invalid input bytes ",
    "(input encoding was '", from_, "')";
```

Reply:
Also, this would be better as `Status::Invalid`, not `Status::IOError`